### PR TITLE
Updates the RoboRioConnection to be class based

### PR DESCRIPTION
**Pull request by ThadHouse**
Will be better for unit testing in the future. Was attempting to fix a
deadlock. This did not fix it, so I think the issue is a bug in the SSH
library. Implemented a workaround.

### CODE REVIEW

File: FRC-Extension/Buttons/KillButton.cs [hunk 1]
```diff
@@ -33,51 +33,37 @@ public KillButton(Frc_ExtensionPackage package)
...
                     await writer.WriteLineAsync("Attempting to Connect to RoboRIO").ConfigureAwait(false);
-
-                    Task<bool> rioConnectionTask = m.StartConnectionTaskAsync(teamNumber);
-                    Task delayTask = Task.Delay(10000);
-
-                    //Successfully extracted files.
-
-                    await writer.WriteLineAsync("Waiting for Connection to Finish").ConfigureAwait(false);
-                    if (await Task.WhenAny(rioConnectionTask, delayTask).ConfigureAwait(false) == rioConnectionTask)
+                    using (var rioConn = await RoboRioConnection.StartConnectionTaskAsync(teamNumber))
```
Comment (333fred) [inline FRC-Extension/Buttons/KillButton.cs]: Missing a `ConfigureAwait`

File: FRC-Extension/RoboRIOCode/DeployManager.cs [hunk 2]
```diff
@@ -39,106 +44,81 @@ public DeployManager(DTE dte)
 
             //Connect to Robot Async
             await OutputWriter.Instance.WriteLineAsync("Attempting to Connect to RoboRIO").ConfigureAwait(false);
-            Task<bool> rioConnectionTask = StartConnectionTaskAsync(teamNumber);
-            Task delayTask = Task.Delay(10000);
+
```
Comment (333fred) [inline FRC-Extension/RoboRIOCode/DeployManager.cs:47]: Whitespace.

File: FRC-Extension/Buttons/KillButton.cs (hunk 1 above)
Comment (ThadHouse) [reply FRC-Extension/Buttons/KillButton.cs]: Fixed

File: FRC-Extension/RoboRIOCode/DeployManager.cs (hunk 2 above)
Comment (ThadHouse) [reply FRC-Extension/RoboRIOCode/DeployManager.cs:47]: Fixed

File: FRC-Extension/RoboRIOCode/DeployManager.cs [hunk 3]
```diff
@@ -39,106 +44,81 @@ public DeployManager(DTE dte)
...
+            Task<RoboRioConnection> rioConnectionTask = RoboRioConnection.StartConnectionTaskAsync(teamNumber);
 
             CodeReturnStruct codeReturn = await BuildAndPrepareCodeAsync(debug, robotProject).ConfigureAwait(false);
 
             if (codeReturn == null) return false;
 
             await writer.WriteLineAsync("Waiting for Connection to Finish").ConfigureAwait(false);
-            if (await Task.WhenAny(rioConnectionTask, delayTask).ConfigureAwait(false) == rioConnectionTask)
+            m_roboRioConnection = await rioConnectionTask.ConfigureAwait(false);
```
Comment (333fred) [inline FRC-Extension/RoboRIOCode/DeployManager.cs:57]: Should be in a using or properly disposed of.

Comment (333fred) [reply FRC-Extension/RoboRIOCode/DeployManager.cs:57]: Wait, nvm. I see

Comment (333fred) [reply FRC-Extension/RoboRIOCode/DeployManager.cs:57]: What about the existing `m_roboRioConnection`? There should be one if code was deployed multiple times.

Comment (ThadHouse) [reply FRC-Extension/RoboRIOCode/DeployManager.cs:57]: Dispose disposes it. That's why using was added to all usages of DeployManager class. I could make it in a using statement in the function, but then I would have to pass it manually to the rest of the functions later.

Comment (ThadHouse) [reply FRC-Extension/RoboRIOCode/DeployManager.cs:57]: Good point about if it exists. I'll fix that.

Comment (ThadHouse) [reply FRC-Extension/RoboRIOCode/DeployManager.cs:57]: Fixed

File: FRC-Extension/RoboRIOCode/DeployManager.cs [hunk 4]
```diff
@@ -250,19 +230,19 @@ public CodeReturnStruct(string exe, List<string> files)
         public async Task<bool> DeployRobotFilesAsync(List<string> files)
         {
             await OutputWriter.Instance.WriteLineAsync("Deploying robot files").ConfigureAwait(false);
-            return await RoboRIOConnection.DeployFiles(files, DeployProperties.DeployDir, ConnectionUser.LvUser).ConfigureAwait(false);
+            return await m_roboRioConnection.DeployFiles(files, DeployProperties.DeployDir, ConnectionUser.LvUser).ConfigureAwait(false);
```
Comment (333fred) [inline FRC-Extension/RoboRIOCode/DeployManager.cs:233]: You're blindly assuming that `m_roboRioConnection` isn't null. I'd suggest making that an assertion, to ensure it actually causes a noticeable error if it ever happens.

Comment (333fred) [reply FRC-Extension/RoboRIOCode/DeployManager.cs:233]: Same with all the ones below.

File: FRC-Extension/RoboRIOCode/RoboRIOConnection.cs [hunk 5]
```diff
@@ -26,73 +30,189 @@ public enum ConnectionUser
...
+            m_sshUserClient?.Dispose();
+            m_scpUserClient?.Dispose();
+            m_scpAdminClient?.Dispose();
+        }
 
+        public static async Task<RoboRioConnection> StartConnectionTaskAsync(string teamNumberS)
+        {
+            int teamNumber;
+            int.TryParse(teamNumberS, out teamNumber);
             if (teamNumber < 0)
```
Comment (333fred) [inline FRC-Extension/RoboRIOCode/RoboRIOConnection.cs:102]: Not changed by your PR, but you should probably throw an exception here. This is bad user input, defaulting to 0 probably isn't going to give a good error.

Comment (ThadHouse) [reply FRC-Extension/RoboRIOCode/RoboRIOConnection.cs:102]: There are other parts that check if it is 0, but that's a good point. I'll do that.

File: FRC-Extension/RoboRIOCode/RoboRIOConnection.cs [hunk 6]
```diff
@@ -26,73 +30,189 @@ public enum ConnectionUser
...
+                            m_connectionType = ConnectionType.LAN;
+                        }
+                        else if (finished == ip)
+                        {
+                            foundHost = ipClient;
+                            m_connectionType = ConnectionType.IP;
+                        }
+                        else
+                        {
+                            // Error
```
Comment (333fred) [inline FRC-Extension/RoboRIOCode/RoboRIOConnection.cs:186]: Again, this should be an impossible situation. Make this easier on yourself to find and use an assert.

File: FRC-Extension/RoboRIOCode/RoboRIOConnection.cs [hunk 7]
```diff
@@ -26,73 +30,189 @@ public enum ConnectionUser
...
+                            return false;
+                        }
+
+                        var ep = foundHost.Client.RemoteEndPoint;
+                        var ipEp = ep as IPEndPoint;
+                        if (ipEp == null)
+                        {
+                            // Continue, we cannot use this one
```
Comment (333fred) [inline FRC-Extension/RoboRIOCode/RoboRIOConnection.cs:194]: Is there ever a realistic situation where `RemoteEndPoint` could be not an `IPEndPoint`? If not, then make this a standard cast, that way an exception will get thrown if this isn't true.

Comment (333fred) [reply FRC-Extension/RoboRIOCode/RoboRIOConnection.cs:194]: If it is possible, invert the if and remove the empty case.

File: FRC-Extension/RoboRIOCode/RoboRIOConnection.cs (hunk 6 above)
Comment (ThadHouse) [reply FRC-Extension/RoboRIOCode/RoboRIOConnection.cs:186]: I don't want this causing an exception, as the message there is worse. Isn't that what an assertion would do?

File: FRC-Extension/RoboRIOCode/RoboRIOConnection.cs [hunk 8]
```diff
@@ -26,73 +30,189 @@ public enum ConnectionUser
...
+                            {
+                                m_remoteIp = ipEp.Address;
+                            }
+                            return finishedConnect;
+                        }
+                    }
+                    tasks.Remove(finished);
+                }
+                // If we have ever gotten here, return false
+                return false;
```
Comment (333fred) [inline FRC-Extension/RoboRIOCode/RoboRIOConnection.cs:210]: Unreachable code, assert.

Comment (ThadHouse) [reply FRC-Extension/RoboRIOCode/RoboRIOConnection.cs:210]: Will do in a separate PR.

File: FRC-Extension/RoboRIOCode/RoboRIOConnection.cs (hunk 7 above)
Comment (ThadHouse) [reply FRC-Extension/RoboRIOCode/RoboRIOConnection.cs:194]: Will fix in a separate PR

File: FRC-Extension/RoboRIOCode/RoboRIOConnection.cs (hunk 6 above)
Comment (ThadHouse) [reply FRC-Extension/RoboRIOCode/RoboRIOConnection.cs:186]: Will fix in a separate PR

File: FRC-Extension/RoboRIOCode/RoboRIOConnection.cs (hunk 5 above)
Comment (ThadHouse) [reply FRC-Extension/RoboRIOCode/RoboRIOConnection.cs:102]: Will fix in a separate PR

File: FRC-Extension/RoboRIOCode/DeployManager.cs (hunk 4 above)
Comment (ThadHouse) [reply FRC-Extension/RoboRIOCode/DeployManager.cs:233]: Will fix this with the assertion PR.

File: FRC-Extension/RoboRIOCode/RoboRIOConnection.cs [hunk 9]
```diff
@@ -123,118 +243,139 @@ public static class RoboRIOConnection
...
-        public static async Task<SshCommand> RunCommandAsync(string command, ConnectionUser user)
+        public async Task<SshCommand> RunCommandAsync(string command, ConnectionUser user)
         {
-            ConnectionInfo connectionInfo;
+            SshClient ssh;
             switch (user)
             {
                 case ConnectionUser.Admin:
-                    connectionInfo = s_adminConnectionInfo;
+                    ssh = m_sshAdminClient;
```
Comment (333fred) [inline FRC-Extension/RoboRIOCode/RoboRIOConnection.cs:352]: Another instance of making sure that you're initialized before use. In the new project system, we have a base class called `OnceInitializedOnceDisposed` that helps to enforce invariants by adding some general purpose initialization blocks, and a method called `EnsureInitialized` that takes a bool, whether to allow initialization now, or throw if not initialized. Something similar might be useful to you.

**Review by 333fred — APPROVED** [review]

File: FRC-Extension/RoboRIOCode/RoboRIOConnection.cs (hunk 9 above)
Comment (ThadHouse) [reply FRC-Extension/RoboRIOCode/RoboRIOConnection.cs:352]: Yeah I'll implement something like that. That's definitely a good idea. Thats not public to be used, right?

Comment (333fred) [reply FRC-Extension/RoboRIOCode/RoboRIOConnection.cs:352]: Well, it's in CPS, so I can't send you the code unfortunately. Once we release the new project system and I bug you into rewriting for that, then you'll be able to use it :).

Comment (ThadHouse) [reply FRC-Extension/RoboRIOCode/RoboRIOConnection.cs:352]: Ha. But that class will be public? That'll be nice. I'll definitely rewrite to use the new project system. Will have to fix the native libraries, but that shouldn't be too bad, and should be a lot cleaner.

Comment (333fred) [reply FRC-Extension/RoboRIOCode/RoboRIOConnection.cs:352]: The implementation won't be public, but that shouldn't stop you from deriving from it, so long long as you add the CPS NuGet package.

**Pull request merged**